### PR TITLE
Getting Gerth to buy keys/misc in iwd+how. (#274)

### PR DESCRIPTION
**Pull request by gregariousmonk**
## Description
First look at enabling Gerth's shop for misc and key items (#274).

Only tested on iwd+how.

[cinnamon-2023-07-11T222144-0230.webm](https://github.com/gemrb/gemrb/assets/137374871/99bfdf10-a8ca-43dd-99f7-82fb67a70b14)

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] I used the same coding style as the surrounding code 
- [x] I have tested the proposed changes
- [ ] I extended the documentation, if necessary
- [ ] The proposed change builds also on our build bots (check after submission)

### CODE REVIEW

**Comment by lynxlynxlynx** [comment]
This is way too crude, enabling selling too often and ignoring the non-sellable bit.

File: gemrb/core/Store.cpp [hunk 1]
```diff
@@ -135,6 +135,15 @@ int Store::AcceptableItemType(ieDword type, ieDword invflags, bool pc) const
 			ret &= ~IE_STORE_SELL;
 		}
 
+		//...the shop is special (iwd/how/bg1)
+		if ((core->HasFeature(GFFlags::NO_UNDROPPABLE)) || (core->HasFeature(GFFlags::BREAKABLE_WEAPONS))) {
+			if (invflags&IE_INV_ITEM_CONVERSABLE) {
+				ret &= ~IE_STORE_SELL;
+			} else {
+				ret |= IE_STORE_SELL;
```
Comment (lynxlynxlynx) [inline gemrb/core/Store.cpp:143]: this could potentially enabling the sale to stores that don't allow it.

Comment (lynxlynxlynx) [reply gemrb/core/Store.cpp:143]: also how should IE_INV_ITEM_DESTRUCTIBLE be handled in this case? Now you reenable their sellability.

Comment (gregariousmonk) [reply gemrb/core/Store.cpp:143]: I added more exact gametype detection.

I'm guessing that the major concern here is getting new/modded games, outside the intended ones, caught in the selling factors?

**Review by gregariousmonk — COMMENTED** [review]
For issue #274

**Review by gregariousmonk — COMMENTED** [review]
For issue #274

File: gemrb/core/Store.cpp [hunk 2]
```diff
@@ -136,7 +136,7 @@ int Store::AcceptableItemType(ieDword type, ieDword invflags, bool pc) const
 		}
 
 		//...the shop is special (iwd/how/bg1)
-		if ((core->HasFeature(GFFlags::NO_UNDROPPABLE)) || (core->HasFeature(GFFlags::BREAKABLE_WEAPONS))) {
+		if ((core->config.GameType == "how") || (core->config.GameType == "bg1")) {
```
Comment (lynxlynxlynx) [inline gemrb/core/Store.cpp:139]: No, like I said in the issue, the proper way is to add another game flag.

File: gemrb/core/Store.cpp (hunk 1 above)
Comment (lynxlynxlynx) [reply gemrb/core/Store.cpp:143]: The gametype checks are there to ensure this code is only used for the games that need it. My comments in this thread are about the code itself — look at the rest of the function, potential execution flow.

File: gemrb/core/Store.cpp (hunk 2 above)
Comment (gregariousmonk) [reply gemrb/core/Store.cpp]: I've added a new game flag and performed a simple test.

[cinnamon-2023-07-13T164020-0230.webm](https://github.com/gemrb/gemrb/assets/137374871/a41e6ec0-c9e9-461d-b2a5-208a61dab0b4)

File: gemrb/unhardcoded/iwd/gemrb.ini [hunk 3]
```diff
@@ -143,6 +143,7 @@ ReverseDoor = 0
 ReverseToHit = 1
 SaveForHalfDamage = 0
 SelectiveMagicRes = 0
+SellableCritsNoConv = 1
```
Comment (lynxlynxlynx) [inline gemrb/unhardcoded/iwd/gemrb.ini:146]: You need to add these also to the other gemrb.ini files, just with 0.

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: Thanks, Jaka, I've pushed the files I missed.

Comment (lynxlynxlynx) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: There's one more, for the demo, just elsewhere in the tree.

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: Fixed and pushed.

Did I miss anything else?

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: Wait, I screwed that one up, sorry.

Comment (lynxlynxlynx) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: The stuff in my first two comments is still not addressed, yeah.

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: I'll sort it out.  Still tracing through the code.

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: Should I make another game flag to enable the selling of items which originally have no bits set for the destructible flag?

Comment (lynxlynxlynx) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: Why, is behaviour different between the games?

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: I can't find where it's being set, but the "critical" items appear to also have the "destructible" bit set to false, the destructible flag doesn't exist in the original iwd/bg1.

If I add a condition for a !"destructible" flag after the condition for critical/not conversable it won't permit the sale.  If I don't add the condition it will permit the sale of indestructible items.

Comment (lynxlynxlynx) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: It's in Interface::SanitizeItem. But again, until the behaviour is proven to be different among games, a game flag makes no sense. There's already a check in this function, I'm just asking what the proper order of the two is, so we can match the originals.

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: Thanks for the leg up, Jaka. I'll go read and break some more code.

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: btw, I agree about the gameflag, that was just frustration talking, sorry.

I have a single test case working with modified interface.cpp code in IWD, I'll unpack bg2 and pst over the weekend and try them out.  

It's been too long since I played pst, too bad it's just for bughunting this time, but it'll be nice to see it again. The GemRB code runs really well, even runs smoothly on a crappy OLPC notebook I use.

File: gemrb/core/Store.cpp [hunk 4]
```diff
@@ -125,27 +125,25 @@ int Store::AcceptableItemType(ieDword type, ieDword invflags, bool pc) const
 		ret &= ~IE_STORE_BUY;
 
 	if (pc && Type < StoreType::BG2CONT) {
-		//don't allow selling of non destructible items
-		if (!(invflags&IE_INV_ITEM_DESTRUCTIBLE )) {
-			ret &= ~IE_STORE_SELL;
-		}
```
Comment (lynxlynxlynx) [inline gemrb/core/Store.cpp:131]: Err, why did you remove this?

File: gemrb/core/Store.cpp [hunk 5]
```diff
@@ -125,27 +125,25 @@ int Store::AcceptableItemType(ieDword type, ieDword invflags, bool pc) const
...
 
 		//don't allow selling of critical items (they could still be put in bags) ... unless the shop is special
-		if ((invflags&IE_INV_ITEM_CRITICAL) && !(Flags&IE_STORE_BUYCRITS)) {
+		if ((invflags&IE_INV_ITEM_CRITICAL) && (!(Flags&IE_STORE_BUYCRITS))) {
```
Comment (lynxlynxlynx) [inline gemrb/core/Store.cpp:130]: This doesn't change anything. Also I'm not sure why you're adding so many parentheses everywhere. Two pairs are enough in this line.

File: gemrb/core/Interface.cpp [hunk 6]
```diff
@@ -3563,6 +3563,8 @@ void Interface::SanitizeItem(CREItem *item) const
 	// some slot flags might be affected by the item flags
 	if (!(item->Flags & IE_INV_ITEM_CRITICAL)) {
 		item->Flags |= IE_INV_ITEM_DESTRUCTIBLE;
+	} else if ((core->HasFeature(GFFlags::SELLABLE_CRITS_NO_CONV)) && (!(item->Flags & IE_INV_ITEM_CONVERSABLE))) {
+		item->Flags |= IE_INV_ITEM_DESTRUCTIBLE;
```
Comment (lynxlynxlynx) [inline gemrb/core/Interface.cpp:3567]: This is not correct. Both bits may be set, so if the critical one isn't, this else-if would get skipped.

File: gemrb/unhardcoded/iwd/gemrb.ini (hunk 3 above)
Comment (lynxlynxlynx) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: No need to test pst stores, as they're not really different and hard to find. It's also almost pointless to test with gemrb, the main questions can only be answered by the originals.

Comment (gregariousmonk) [reply gemrb/unhardcoded/iwd/gemrb.ini:146]: Gorion's  Scroll (scrl3z) has the critical flag and can be sold to Bentley Mirrorshade in the original (non-EE) bg+totsc.

I'll continue testing...

File: gemrb/core/Interface.cpp [hunk 7]
```diff
@@ -3563,10 +3563,13 @@ void Interface::SanitizeItem(CREItem *item) const
 	// some slot flags might be affected by the item flags
 	if (!(item->Flags & IE_INV_ITEM_CRITICAL)) {
 		item->Flags |= IE_INV_ITEM_DESTRUCTIBLE;
-	} else if ((core->HasFeature(GFFlags::SELLABLE_CRITS_NO_CONV)) && (!(item->Flags & IE_INV_ITEM_CONVERSABLE))) {
+	} else if (core->HasFeature(GFFlags::SELLABLE_CRITS_NO_CONV)) {
 		item->Flags |= IE_INV_ITEM_DESTRUCTIBLE;
-	}
-
+		}
+	if  (item->Flags & IE_INV_ITEM_CONVERSABLE) {
```
Comment (lynxlynxlynx) [inline gemrb/core/Interface.cpp:3569]: This needs to happen only in iwd, unless you have some evidence items with the conversable bit shouldn't be destructible.

Comment (lynxlynxlynx) [reply gemrb/core/Interface.cpp:3569]: And to quote the expected behaviour:
> All items survive effects like disintegration and freezing/stone death irrespective of bit 0

IOW no items should have the bit set.

Comment (gregariousmonk) [reply gemrb/core/Interface.cpp:3569]: I couldn't find any items with bit 11 set in vanilla totsc, so iwd appears to be the only one using it despite the comment in the ITM docs.  I have yet to test/compare with bg2.

From the behaviour of the original games a 'critical' bit should permit selling in both bg1 and iwd/how, but the 'conversable' bit only appears to be used as unsellable in iwd/how.  

Should this be two separate flags as well or stick with one flag?

Comment (lynxlynxlynx) [reply gemrb/core/Interface.cpp:3569]: Isn't this still the same difference, already documented in the initial report? 

I can't think of any users of bit 11 outside of bg2 (lilarcor) off the top of my head. The chatty sword in iwd you can't interact with and I'm not sure if pst needs it, since it has a whole field for the dialog name.

File: gemrb/core/Interface.cpp [hunk 8]
```diff
@@ -3566,7 +3566,7 @@ void Interface::SanitizeItem(CREItem *item) const
 	} else if (core->HasFeature(GFFlags::SELLABLE_CRITS_NO_CONV)) {
 		item->Flags |= IE_INV_ITEM_DESTRUCTIBLE;
 		}
-	if  (item->Flags & IE_INV_ITEM_CONVERSABLE && core->HasFeature(GFFlags::SELLABLE_CRITS_NO_CONV)) {
+	if  ((item->Flags & IE_INV_ITEM_CONVERSABLE) && core->HasFeature(GFFlags::SELLABLE_CRITS_NO_CONV)) {
```
Comment (lynxlynxlynx) [inline gemrb/core/Interface.cpp:3569]: useless change: https://en.cppreference.com/w/cpp/language/operator_precedence

**Comment by lynxlynxlynx** [comment]
Ok, took a fresh view:
- the demo is still missing the new key in its ini
- you're still enabling the sale for shops that don't support it (eg. temples), but
- with the changes to Interface.cpp, does Store.cpp even need changing at all?

**Comment by gregariousmonk** [comment]
The change to the test gemrb.ini is in commit aba64f8, I'm not sure why it's not showing up here.

I'll look at the code for interface and store, but I believe you may be correct, the store code should not be needed anylonger since the filtering would happen in interface.  I'll check it this week.

**Comment by lynxlynxlynx** [comment]
Hey, any progress on this?

**Comment by lynxlynxlynx** [comment]
Do you plan to finish this or should we take over?

**Comment by gregariousmonk** [comment]
Hi Jaka, I will finish it.  Apologies for the delay, I've had a large
amount of overtime at work for the past few months and am currently tending
to my wife in hospital in Korea.  I return home in a couple of weeks and
will look at it then.

**Comment by lynxlynxlynx** [comment]
I've completed your work and pushed it; it will be available to everyone with the next release. Thanks!
